Change related entity with PUT (#1994)
oasis-tcs/odata-specs: odata-protocol/11.4 Data Modification.md

set of entities to be related according to that relationship and MUST
NOT include added links, deleted links, or deleted entities.

If a navigation property is absent from a `PUT` or `PATCH` request payload, the referenced

> or contained entity, or the collection thereof, which the navigation property targets

Reviewer: ```suggestion
or contained entity, or the collection thereof,
```

  ]
}
```
:::

::: example
Example ##ex: When updating an entity with a 4.01 `PUT` request, the target of a
non-containment navigation property can be replaced without changing either the
old or the new targeted entity itself. The targeted entity can then only be

> specified by an entity reference (see [OData-JSON, section "Entity Reference"]($$$OData-JSON$$$#EntityReference)):

Reviewer: "specified as"?

```json
PUT http://host/service/Employees(7)
Content-Type: application/json

{
  "FirstName": "Suzanne",
  "LastName": "Brown-Johnson",
  "Manager": {
    "@context": "$metadata#$ref",
    "@id": "Employees(6)"
  }
}
```

If the targeted entity in the payload contains some structural properties,
`PUT` resets all its other structural properties. The following therefore does
not have the desired effect:
```json
PUT http://host/service/Employees(7)
Content-Type: application/json

{
  "FirstName": "Suzanne",
  "LastName": "Brown-Johnson",
  "Manager": {

>     "@id": "Employees(6)",
>     "EmployeeID": 6

Reviewer: ```suggestion
    "EmployeeID": 6
```

Author: We should clarify both alternatives:
* reference plus some properties
* no reference, just some properties.

> If the targeted entity in the payload contains some structural properties,
> `PUT` resets all its other structural properties. The following therefore does
> not have the desired effect:
> ```json
> PUT http://host/service/Employees(7)
> Content-Type: application/json
> 
> {
>   "FirstName": "Suzanne",
>   "LastName": "Brown-Johnson",
>   "Manager": {
>     "@id": "Employees(6)",
>     "EmployeeID": 6
>   }
> }

Author: Or even stricter?
```suggestion
If the targeted entity in the payload does not have the `@context` of an entity reference,
`PUT` resets all its structural properties not contained in the payload. The following therefore does
not have the desired effect:
```json
PUT http://host/service/Employees(7)
Content-Type: application/json

{
  "FirstName": "Suzanne",
  "LastName": "Brown-Johnson",
  "Manager": {
    "@id": "Employees(6)"
  }
}
```

Reviewer: That's too strict: an object only containing `@id` essentially is an entity reference, even without the `@context`.

Reviewer: I'd rather remove the `@id` from the "undesired effect" example

```

If the targeted entity in the payload contains some structural properties,
`PUT` resets all its other structural properties. The following alternative
payloads both reset the first and last names of the new manager:
```json
{
  "FirstName": "Suzanne",
  "LastName": "Brown-Johnson",
  "Manager": {
    "@context": "$metadata#$ref",
    "@id": "Employees(6)",

>     "EmployeeID": 6

Reviewer: That is not a valid entity reference, additional properties are not allowed.

Reviewer: If you remove the `@context`, it becomes an incomplete entity and a valid example for undesired effects.
